tooling/ssa_fuzzer/fuzzer/src/mutations/mod.rs

```diff
@@ -1,36 +1,90 @@
+    BASIC_FUNCTION_VEC_MUTATION_CONFIGURATION, FunctionVecMutationOptions,
 };
+use libfuzzer_sys::arbitrary::Unstructured;
 use rand::{Rng, rngs::StdRng};
 
-pub(crate) fn mutate(data: FuzzerData, rng: &mut StdRng) -> FuzzerData {
-    let (mut blocks, mut commands, mut initial_witness) =
-        (data.blocks, data.commands, data.initial_witness);
-    let number_of_mutations = rng.gen_range(1..MAX_NUMBER_OF_MUTATIONS);
-    for _ in 0..number_of_mutations {
-        match BASIC_MUTATION_CONFIGURATION.select(rng) {
-            MutationOptions::InstructionBlocks => {
-                instructions::mutate_vec_instruction_block(&mut blocks, rng);
-            }
-            MutationOptions::FuzzerCommands => {
-                commands_mutator::mutate_vec_fuzzer_command(&mut commands, rng);
-            }
-            MutationOptions::Witnesses => {
-                let index = rng.gen_range(0..initial_witness.len());
-                witness_mutator::witness_mutate(&mut initial_witness[index], rng);
-            }
+trait MutateFunctionVec {
+    fn mutate(rng: &mut StdRng, funcs: &mut Vec<FunctionData>);
+}
+struct RandomMutation;
+impl MutateFunctionVec for RandomMutation {
+    fn mutate(rng: &mut StdRng, functions: &mut Vec<FunctionData>) {
+        let mut bytes = [0u8; 128];
+        rng.fill(&mut bytes);
+        *functions = Unstructured::new(&bytes).arbitrary().unwrap();
+    }
+}
+
+struct RemoveFunctionMutation;
+impl MutateFunctionVec for RemoveFunctionMutation {
+    fn mutate(rng: &mut StdRng, functions: &mut Vec<FunctionData>) {
+        if !functions.is_empty() {
+            functions.remove(rng.gen_range(0..functions.len()));
+        }
+    }
+}
+
+/// Replace randomly chosen function with randomly generated function
+struct ReplaceFunctionMutation;
+impl MutateFunctionVec for ReplaceFunctionMutation {
+    fn mutate(rng: &mut StdRng, functions: &mut Vec<FunctionData>) {
+        let mut bytes = [0u8; 25];
+        rng.fill(&mut bytes);
+        let command = Unstructured::new(&bytes).arbitrary().unwrap();
+        if !functions.is_empty() {
+            let command_idx = rng.gen_range(0..functions.len());
+            functions[command_idx] = command;
+        }
+    }
+}
+
+struct MutateFunction;
+impl MutateFunctionVec for MutateFunction {
+    fn mutate(rng: &mut StdRng, functions: &mut Vec<FunctionData>) {
+        if functions.is_empty() {
+            return;
+        }
+        let function_idx = rng.gen_range(0..functions.len());
+        function::mutate_function(&mut functions[function_idx], rng);
+    }
+}
+
+/// Insert randomly generated function
+struct FunctionInsertionMutation;
+impl MutateFunctionVec for FunctionInsertionMutation {
+    fn mutate(rng: &mut StdRng, functions: &mut Vec<FunctionData>) {
+        let mut bytes = [0u8; 25];
+        rng.fill(&mut bytes);
+        let function = Unstructured::new(&bytes).arbitrary().unwrap();
```

Comment:
A fully random function is probably too random. You want to create stepping stones for your fuzzer